sdk/eventhub/event-hubs/src/partitionLoadBalancer.ts

```diff
@@ -0,0 +1,358 @@
+    partitionIdToClaim: string
+  ): Promise<void> {
+    log.partitionLoadBalancer(
+      `[${this._ownerId}] Attempting to claim ownership of partition ${partitionIdToClaim}.`
+    );
+    const ownershipRequest = this._createPartitionOwnershipRequest(
+      partitionOwnershipMap,
+      partitionIdToClaim
+    );
+    try {
+      await this._partitionManager.claimOwnership([ownershipRequest]);
+      log.partitionLoadBalancer(
+        `[${this._ownerId}] Successfully claimed ownership of partition ${partitionIdToClaim}.`
+      );
+      const partitionContext: PartitionContext = {
+        consumerGroupName: this._consumerGroupName,
+        eventHubName: this._eventHubClient.eventHubName,
+        partitionId: ownershipRequest.partitionId
+      };
+
+      const checkpointManager = new CheckpointManager(
+        partitionContext,
+        this._partitionManager,
+        this._ownerId
+      );
+
+      log.partitionLoadBalancer(
+        `[${this._ownerId}] [${partitionIdToClaim}] Calling user-provided PartitionProcessorFactory.`
+      );
+      const partitionProcessor = this._partitionProcessorFactory(
+        partitionContext,
+        checkpointManager
+      );
+
+      const eventPosition = ownershipRequest.sequenceNumber
+        ? EventPosition.fromSequenceNumber(ownershipRequest.sequenceNumber)
+        : this._processorOptions.initialEventPosition || EventPosition.earliest();
+
+      await this._pumpManager.createPump(
+        this._eventHubClient,
+        partitionContext,
+        eventPosition,
+        partitionProcessor
+      );
+      log.partitionLoadBalancer(`[${this._ownerId}] PartitionPump created successfully.`);
+    } catch (err) {
+      log.error(
+        `[${this._ownerId}] Failed to claim ownership of partition ${ownershipRequest.partitionId}`
+      );
+    }
+  }
+
+  /*
+   * This method is called after determining that the load is not balanced. This method will evaluate
+   * if the current event processor should own more partitions. Specifically, this method returns true if the
+   * current event processor owns less than the minimum number of partitions or if it owns the minimum number
+   * and no other event processor owns lesser number of partitions than this event processor.
+   */
+  private _shouldOwnMorePartitions(
+    minPartitionsPerEventProcessor: number,
+    ownerPartitionMap: Map<string, PartitionOwnership[]>
+  ): boolean {
+    const numberOfPartitionsOwned = ownerPartitionMap.get(this._ownerId)!.length;
+
+    let leastPartitionsOwnedByAnyEventProcessor = Number.MAX_VALUE;
+    for (const ownershipList of ownerPartitionMap.values()) {
+      if (ownershipList.length < leastPartitionsOwnedByAnyEventProcessor) {
+        leastPartitionsOwnedByAnyEventProcessor = ownershipList.length;
+      }
+    }
+    return (
+      numberOfPartitionsOwned < minPartitionsPerEventProcessor ||
+      numberOfPartitionsOwned === leastPartitionsOwnedByAnyEventProcessor
```

Comment:
We won't reach this code if the load is already balanced. On line 185, there's a check to see if the load is already balanced. This method gets called only when the load is not balanced.

Reply:
That's true, but I still think it was right to change the logic since the function should still act as expected in isolation. (If we changed the code such that we didn't check _isLoadBalanced first this method should still return the expected result)

Reply:
I think what you said is true only if `_shouldOwnMorePartitions()` was public, but I don't think that if the method is private then it has to work in isolation. If anyone will change the code, they should check the logic before using private methods in a class.